internal/services/v1/debug.go

```diff
@@ -0,0 +1,188 @@
+package v1
+
+import (
+	"context"
+	"strings"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	cexpr "github.com/authzed/spicedb/internal/caveats"
+	"github.com/authzed/spicedb/pkg/datastore"
+	dispatch "github.com/authzed/spicedb/pkg/proto/dispatch/v1"
+	"github.com/authzed/spicedb/pkg/schemadsl/compiler"
+	"github.com/authzed/spicedb/pkg/schemadsl/generator"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// convertDispatchDebugInformation converts dispatch debug information found in the response metadata
+// into DebugInformation returnable to the API.
+func convertDispatchDebugInformation(
+	ctx context.Context,
+	caveatContext map[string]any,
+	metadata *dispatch.ResponseMeta,
+	reader datastore.Reader,
+) (*v1.DebugInformation, error) {
+	debugInfo := metadata.DebugInfo
+	if debugInfo == nil {
+		return nil, nil
+	}
+
+	caveats, err := reader.ListCaveats(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	namespaces, err := reader.ListNamespaces(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defs := make([]compiler.SchemaDefinition, 0, len(namespaces)+len(caveats))
+	for _, caveat := range caveats {
+		defs = append(defs, caveat)
+	}
+	for _, ns := range namespaces {
+		defs = append(defs, ns)
+	}
+
+	schema, _ := generator.GenerateSchema(defs)
+	converted, err := convertCheckTrace(ctx, caveatContext, debugInfo.Check, reader)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.DebugInformation{
+		Check:      converted[0],
```

Comment:
I assume `LookupResources` does not have debug support yet. In the meantime, we can assume `dispatch.CheckDebugTrace` is going to have only information for one checked resource, and then change `convertCheckTrace` to return one single  `v1.CheckDebugTrace` to make clear these are only used for `Check`-initiated dispatches.

Or we can make it more explicit in the method name `convertDispatchDebugInformation` that there is an expectation the dispatch debug information comes from a Check-initiated dispatch and thus has one single resource.

Reply:
Renamed to `convertCheckDispatchDebugInformation`